### PR TITLE
Add custom material processing and pipeline type for pipeline.

### PR DESCRIPTION
**Pull request by caryliu1999**
https://github.com/cocos-creator/engine-native/pull/3716

### CODE REVIEW

File: cocos/core/pipeline/pipeline-scene-data.ts [hunk 1]
```diff
@@ -61,10 +62,30 @@ export class PipelineSceneData {
...
+        if (this._deferredLightingMat === mat) return;
+        this._deferredLightingMat = (mat && mat.effectAsset) ? mat : builtinResMgr.get<Material>('builtin-deferred-material');
+        this.updateDeferredPassInfo();
+    }
+
+    public get deferredPostMat () {
+        return this._deferredPostMat;
+    }
+
+    public set deferredPostMat (mat: Material) {
```
Comment (pandamicro) [inline cocos/core/pipeline/pipeline-scene-data.ts]: The above api naming don't follow our rules, Mat should be Material instead

File: cocos/core/pipeline/define.ts [hunk 2]
```diff
@@ -63,6 +63,15 @@ export enum RenderPriority {
     DEFAULT = 0x80,
 }
 
+/**
+ * @en The pipeline render type.
+ * @zh 渲染管线类型。
+ */
+export enum PipelineType {
```
Comment (pandamicro) [inline cocos/core/pipeline/define.ts]: The pipeline is designed for customization, this enum just make it static. I think the pipeline scene data is a future obstacle for customized pipeline, the different operations should be placed into the pipeline logic. If it's hard to modify currently, I suggest to at least make some APIs private:

- This enum
- pipelineType in Game
- deferred related properties in pipeline-scene-data

I think we can extend PipelineSceneData in each pipeline

File: cocos/core/pipeline/deferred/postprocess-stage.ts [hunk 3]
```diff
@@ -167,25 +158,18 @@ export class PostprocessStage extends RenderStage {
...
-        if (builtinPostProcess) {
-            pass = builtinPostProcess.passes[0];
-            shader = ShaderPool.get(pass.getShaderVariant());
-        } else {
-            pass = this._postprocessMaterial!.passes[POSTPROCESSPASS_INDEX];
-            shader = ShaderPool.get(this._postprocessMaterial!.passes[POSTPROCESSPASS_INDEX].getShaderVariant());
-        }
+        const builtinPostProcess = sceneData.deferredPostMat;
+        const pass = builtinPostProcess.passes[0];
+        const shader = ShaderPool.get(pass.getShaderVariant());
```
Comment (2youyou2) [inline cocos/core/pipeline/deferred/postprocess-stage.ts]: Postprocess stage should bind material descriptorSet too.

```js
cmdBuff.bindDescriptorSet(SetIndex.MATERIAL, pass.descriptorSet);
```

Comment (caryliu1999) [reply cocos/core/pipeline/deferred/postprocess-stage.ts]: Already added it

File: cocos/core/pipeline/pipeline-scene-data.ts (hunk 1 above)
Comment (caryliu1999) [reply cocos/core/pipeline/pipeline-scene-data.ts]: > The above api naming don't follow our rules, Mat should be Material instead

Already rename all the Mat.

File: cocos/core/pipeline/define.ts (hunk 2 above)
Comment (caryliu1999) [reply cocos/core/pipeline/define.ts]: > The pipeline is designed for customization, this enum just make it static. I think the pipeline scene data is a future obstacle for customized pipeline, the different operations should be placed into the pipeline logic. If it's hard to modify currently, I suggest to at least make some APIs private:
> 
> * This enum
> * pipelineType in Game
> * deferred related properties in pipeline-scene-data
> 
> I think we can extend PipelineSceneData in each pipeline

Already delete enum PipelineType and add deferred-pipeline-scene-data.

**Comment by pandamicro** [comment]
@caryliu1999 please delay this PR to v3.3

**Review by pandamicro — APPROVED** [review]

**Review by minggo — APPROVED** [review]

**Pull request merged**